### DISCUSSION TITLE
`curl_global_cleanup()` needed before exiting binary (exe)?

### DISCUSSION THREAD

**Discussion by user706** [curl/curl#13675] [Q&A] [answered]
Hi,

is `curl_global_cleanup()` needed if I'm exiting the binary (executable)?

In other words: 
What happens if I repeatedly do the following sequentially (say **thousands of times per minute**):

- start an executable that runs `curl_global_init(CURL_GLOBAL_ALL);`  
and then immediately exits _without_ (!) calling `curl_global_cleanup()`.

??

Will I somehow have less and less memory?  Or any other problem?
Or will everything be clean, as if `curl_global_cleanup()` would have been called before exiting?

Background: If it is really needed, then I have to write signal handlers, such that I ensure that `curl_global_cleanup()` is still called when terminating an application/service via `systemctl stop myapplication`. Is that really needed? Any pointers? Thanks.

Thanks.

PS: Will things be somehow different, if I start such an application thousands of times per minute.... **concurrently**, instead of sequentially?

**Comment by dfandrich** [comment, marked as answer]
On Thu, May 16, 2024 at 05:24:49AM -0700, user706 wrote:
> What happens if I repeatedly do the following sequentially (say thousands of
> times per minute):
> 
>   • start an executable that runs curl_global_init(CURL_GLOBAL_ALL);
>     and then immediately exits without (!) calling curl_global_cleanup().
> 
> ??
> 
> Will I somehow have less and less memory? Or any other problem?
> Or will everything be clean, as if curl_global_cleanup() would have been called
> before exiting?

Once a process exits, your operating system will free up all the nonpersistent
resources used by that process, so there won't be anything left once it's done.
Theoretically, some libcurl dependency could leave things like lock files lying
around that wouldn't be cleaned up, but I'm not aware of anything like that
currently being an issue. Also, leak checking systems like valgrind or
LeakSanitizer will raise warnings on exit, but you probably aren't planning on
using those.

**Reply by user706** [reply]
I know all this.

But the curl documentation is really quite weird in that respect, as it seems to suggest that I really **must** call `curl_global_cleanup()` before exiting the program. Just check the example, where it's called before main() exits: https://curl.se/libcurl/c/curl_global_cleanup.html

So I'm wondering... maybe it allocates resources in some other service (and different unrelated process)... maybe openssl or TLS or whatever??

**Comment by dfandrich** [comment]
If you're concerned about cleaning up everything properly in all cases, follow
the documentation and call curl_global_cleanup() before exit. But if your
testing reveals that the OS is doing a fine job of cleaning up on its own, you
always have the freedom to ignore the documentation.
